processing/src/main/java/io/druid/query/topn/TopNLexicographicResultBuilder.java

```diff
@@ -122,9 +183,18 @@ public int compare(
               DimValHolder o2
           )
           {
-            return comparator.compare(o1.getDimName(), o2.getDimName());
+            return comparator.compare(o2.getDimName(), o1.getDimName());
           }
         }
-    ).maximumSize(threshold).create();
+    );
   }
+
+  private boolean shouldAdd(String dimName)
+  {
+    final boolean belowThreshold = pQueue.size() < this.threshold;
+    final boolean belowMax = belowThreshold
+                             || this.comparator.compare(pQueue.peek().getTopNMetricVal(), dimName) < 0;
+    return belowMax && comparator.compare(dimName, previousStop) > 0;
```

Comment:
the implementation in `TopNNumericResultBuilder` doesn't seem to have the added `&& comparator.compare(dimName, previousStop) > 0`. Do you mind adding a comment to explain why it's needed?